problem_builder/tests/unit/test_instructor_tool.py

```diff
@@ -0,0 +1,226 @@
+            {'usage_id': u'3', 'preferred_attr': 'name', 'attrs': {'name': 'name'}},
+            {'usage_id': u'4', 'preferred_attr': 'display_name', 'attrs': {'display_name': 'display_name'}},
+            # Two attributes (question, name): Prefer question
+            {
+                'usage_id': u'5',
+                'preferred_attr':
+                'question', 'attrs': {'question': 'question', 'name': 'name'}
+            },
+            # Two attributes (question, display_name): Prefer question
+            {
+                'usage_id': u'6',
+                'preferred_attr': 'question',
+                'attrs': {'question': 'question', 'display_name': 'display_name'}
+            },
+            # Two attributes (name, display_name): Prefer name
+            {
+                'usage_id': u'7',
+                'preferred_attr': 'name',
+                'attrs': {'name': 'name', 'display_name': 'display_name'}
+            },
+            # All attributes: Prefer question
+            {
+                'usage_id': u'8',
+                'preferred_attr': 'question',
+                'attrs': {'question': 'question', 'name': 'name', 'display_name': 'display_name'}
+            },
+        ]
+
+    def test_student_view_block_tree_names(self):
+        """
+        Check if `student_view` method uses preferred block attributes
+        when building block tree for render context.
+
+        Each entry of a block tree is a dictionary that must contain
+        an `id` key and a `name` key.
+
+        - `id` must be set to the ID
+          (usage_id or block_id) of the corresponding block.
+
+        - `name` must be set to the value of one of the following
+          attributes of the corresponding block:
+
+          - question
+          - name (question ID)
+          - display_name (question title)
+          - block ID
+
+          Note that the attributes are listed in order of preference;
+          i.e., if `block.question` has a meaningful value, that value
+          should be used for `name` (irrespective of what the values
+          of the other attributes might be).
+        """
+        instructor_tool.loader = Mock()
+        instructor_tool.loader.render_template.return_value = u''
+
+        self.block.student_view(context={})
+        block_tree = instructor_tool.loader.render_template.call_args[0][1]['block_tree']
+
+        # Check size of block_tree
+        num_blocks = len(self.block.children) + 1
+        self.assertEqual(len(block_tree), num_blocks)
+
+        # Check block_tree for root entry
+        # - Does block_tree contain single entry whose `id` matches usage_id of instructor block?
+        matching_blocks = [block for block in block_tree if block['id'] == self.block.scope_ids.usage_id]
+        self.assertEqual(len(matching_blocks), 1)
+        # # - Is `name` of that entry set to "All"?
+        matching_block = matching_blocks.pop()
+        self.assertEqual(matching_block['name'], 'All')
```

Comment:
@itsjeyd nit: could use a helper method:

    def check_block(usage_id, expected_name):
        matching_blocks = [block for block in block_tree if block['id'] == self.block.scope_ids.usage_id]
        self.assertEqual(len(matching_blocks), 1)
        matching_block = matching_blocks[0]
        self.assertEqual(matching_block['name'], expected_name)